### PR TITLE
Add eth_signTransaction and eth_sendRawTransaction

### PR DESCRIPTION
**Pull request by darkdh**
Resolves https://github.com/brave/brave-browser/issues/23582
1. Add `eth_sendRawTransaction` which doesn't need confirmation like MetaMask because it doesn't have account actions
2. Add `eth_signTransaction` by adding additional `signOnly` and `signedTransaction` passed along with `eth_sendTransaction` code flow
3. Use idl for transaction request params parsing
4. Introduce a new state "Signed" which indicates the transaction is signed but not submitted. When we confirm that dapps has broadcast it into blocks, it will be moved to "Confirmed" state. "Signed" transaction won't be taken into account when calculating nonce and it will be dropped when its nonce became too low.
5. Confirmation panel is same as same transaction but post confirmation panel is different
Screen Shot 2022-12-01 at 17 53 49

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
>   `eth_signTransaction` is basically same as `eth_sendTransaction` but without sending. And `eth_sendRawTransaction` is just forwarding request to rpc endpoint without account involved
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Normal flow
1. Setup wallet
2. Setup https://github.com/bbondy/eth-manual-tests 
3. Change to any test network and visit http://localhost:8081/request.html
4. Click `ethereum.enable` and grant permission
5. Fill to address next to `eth_signTransaction` and click it
6. Check transactions on brave://wallet/crypto/portfolio/ETH and there should be a "Signed" transaction
7. Go back to the console of http://localhost:8081/request.html
8. Paste the return signed transaction in console to `eth_sendRawTransaction` and click
9. Check transactions on brave://wallet/crypto/portfolio/ETH and previously "Signed" transaction should be "Confirmed" after a while

### Nonce drop
1. Setup wallet
2. Setup https://github.com/bbondy/eth-manual-tests 
3. Change to any test network and visit http://localhost:8081/request.html
4. Click `ethereum.enable` and grant permission
5. Fill to address next to `eth_signTransaction` and click it
6. Fill to address next to `eth_signTransaction (1559)` and click it
7. Check transactions on brave://wallet/crypto/portfolio/ETH and there should be TWO "Signed" transactions
8. Go back to the console of http://localhost:8081/request.html
9. Paste one of the return signed transactions in console to `eth_sendRawTransaction` and click
10. Check transactions on brave://wallet/crypto/portfolio/ETH and previously one "Signed" transaction should be "Confirmed" and one "Signed" transaction should be deleted after a while

### CODE REVIEW

File: android/android_browser_tests.gni [hunk 1]
```diff
@@ -24,7 +24,7 @@ android_test_exception_sources = [
   "//brave/browser/brave_wallet/brave_wallet_sign_message_browsertest.cc",
   "//brave/browser/brave_wallet/brave_wallet_tab_helper_browsertest.cc",
   "//brave/browser/brave_wallet/ethereum_provider_browsertest.cc",
-  "//brave/browser/brave_wallet/send_transaction_browsertest.cc",
```
Comment (wchen342) [inline android/android_browser_tests.gni:27]: If possible can you make them run on Android too? We aim to minimize this list as much as possible so Android will be covered.

**Review by wchen342 — APPROVED** [review]
++

File: components/resources/wallet_strings.grdp [hunk 2]
```diff
@@ -510,6 +511,8 @@
   <message name="IDS_BRAVE_WALLET_ACCOUNT_FILTER_ALL_ACCOUNTS" desc="Account Filter Selector All Accounts">All Accounts</message>
   <message name="IDS_BRAVE_WALLET_TRANSACTION_SUBMITTED_TITLE" desc="Transaction submitted title">Transaction submitted</message>
   <message name="IDS_BRAVE_WALLET_TRANSACTION_SUBMITTED_DESCRIPTION" desc="Transaction submitted description">Transaction has been successfully sent to the network and awaits confirmation.</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_SIGNED_TITLE" desc="Transaction signed title">Transaction signed</message>
+  <message name="IDS_BRAVE_WALLET_TRANSACTION_SIGNED_DESCRIPTION" desc="Transaction signed description">Transaction has been signed and will be sent to network by dapps and awaits confirmation</message>
```
Comment (bbondy) [inline components/resources/wallet_strings.grdp:515]: The transaction has been signed and will be sent to the network by the DApp

File: components/brave_wallet/common/brave_wallet.mojom [hunk 3]
```diff
@@ -982,7 +984,8 @@ enum TransactionStatus {
   Submitted = 3,
   Confirmed = 4,
   Error = 5,
-  Dropped = 6
+  Dropped = 6,
+  Signed = 7 // dapps will submit the transaction
```
Comment (bbondy) [inline components/brave_wallet/common/brave_wallet.mojom:988]: Does this need to be flagged to iOS or Android for handling in any switches?For Android is there anything in this PR that is needed?

Comment (bbondy) [reply components/brave_wallet/common/brave_wallet.mojom:988]: For Desktop, do we need anything in any of these which have other switch cases? 
- `components/brave_wallet_ui/components/shared/style.tsx`
- `components/brave_wallet_ui/components/extension/confirm-transaction-panel/confirm-solana-transaction-panel.tsx`
- `components/brave_wallet_ui/components/desktop/portfolio-transaction-item/index.tsx` (You have one covered in this file but there's another place)

I just did `git grep \\.Dropped` to get this list.

**Review by bbondy — CHANGES_REQUESTED** [review]
Looks good, just some questions and a minor text correction.

File: components/brave_wallet/common/brave_wallet.mojom (hunk 3 above)
Comment (darkdh) [reply components/brave_wallet/common/brave_wallet.mojom:988]: Yes, iOS and Android need to handle the new tx status on UI. cc @SergeyZhukovsky @StephenHeaps.
`git grep TransactionStatus\\.` is more precise so you can see what is being used on Android.
I will check desktop UI for missing parts.

Comment (darkdh) [reply components/brave_wallet/common/brave_wallet.mojom:988]: I don't see other changes needed on UI except `components/brave_wallet_ui/components/shared/style.tsx` to update its opacity

Comment (darkdh) [reply components/brave_wallet/common/brave_wallet.mojom:988]: I will open two issues for iOS and Android as follow-up

**Review by bbondy — APPROVED** [review]

File: components/brave_wallet/common/brave_wallet.mojom (hunk 3 above)
Comment (darkdh) [reply components/brave_wallet/common/brave_wallet.mojom:991]: https://github.com/brave/brave-browser/issues/27224 and https://github.com/brave/brave-ios/issues/6606

**Pull request merged**